### PR TITLE
Save map layer visibility.

### PR DESCRIPTION
**Pull request by mdastous-bentley**
@pmconne What do you think of this?  We want the layer visibility (sub-layer visibility is already fine)  to be saved.  As a side effect, if your visible flag was undefined, then it will now default to 'true' if you do the json roundtrip with MapLayerSettings.  The behaviour is similar to 'formatId' which defaults to 'WMS' if unspecified.

### CODE REVIEW

**Comment by pmconne** [comment]
Seems like an oversight that it wasn't being included by `toJSON`, right?
I get a little grossed out by optional boolean properties that default to `true` (`transparentBackground` also does this).
Can you update the documentation of those two properties to specify they default to true? Currently, the following suggests the opposite:
```
  /** True to indicate background is tranparent */
  transparentBackground?: boolean;
```

**Comment by pmconne** [comment]
Would you mind also figuring out why the MapLayerSettings constructor bypasses the `no-undef-init` eslint rule and ideally make it not have to?

**Comment by mdastous-bentley** [comment]
> Seems like an oversight that it wasn't being included by `toJSON`, right?
> I get a little grossed out by optional boolean properties that default to `true` (`transparentBackground` also does this).
> Can you update the documentation of those two properties to specify they default to true? Currently, the following suggests the opposite:
> 
> ```
>   /** True to indicate background is tranparent */
>   transparentBackground?: boolean;
> ```

Actually its optional in MapLayerProps, but defaults to true when used to initialize MapLayerSettings (which is misleading).  Should we document the defaults in MapLayerProps doc anyway?

**Comment by pmconne** [comment]
I don't understand what distinction you are making. The JSON property is optional; if it is omitted, MapLayerSettings.fromJSON() will default it to true. That should be documented in MapLayerProps (as should all default values).

**Comment by mdastous-bentley** [comment]
I also see maxZoom on MapLayerProps, but missing MapLayerSettings. It  doesnt seem to have any use.

**Comment by pmconne** [comment]
> I also see maxZoom on MapLayerProps, but missing MapLayerSettings. It doesnt seem to have any use.

MapLayerSource.fromJSON() takes its json as `any` (why?!) and pulls a `maxZoom` out of it, among other MapLayerProps properties...

File: core/common/src/MapLayerSettings.ts [hunk 1]
```diff
@@ -109,23 +109,23 @@ export class MapSubLayerSettings {
...
   url?: string;
   /** Source layers. If undefined all layers are displayed. */
   subLayers?: MapSubLayerProps[];
-  /** A transparency value from 0.0 (fully opaque) to 1.0 (fully transparent) to apply to map graphics when drawing, or false to indicate the transparency should not be overridden. Default value: 0. */
+  /** A transparency value from 0.0 (fully opaque) to 1.0 (fully transparent) to apply to map graphics when drawing,
+   * or false to indicate the transparency should not be overridden. Default value: 0.
+   * If omitted, defaults to 0 in MapLayerSettings.fromJSON() */
   transparency?: number;
-  /** True to indicate background is tranparent */
+  /** True to indicate background is transparent.  If omitted, defaults to 'true' in MapLayerSettings.fromJSON() */
```
Comment (pmconne) [inline core/common/src/MapLayerSettings.ts:126]: It also defaults to true in MapLayerSource.fromJSON(), no? I would just say it defaults to true.

File: core/frontend/src/tile/map/MapLayerSources.ts [hunk 2]
```diff
@@ -25,16 +25,37 @@ export enum MapLayerSourceStatus {
   RequireAuth,
 }
 
+/** JSON representation of a map layer source.
+ * @internal
+ */
+interface MapLayerSourceProps extends MapLayerProps
```
Comment (pmconne) [inline core/frontend/src/tile/map/MapLayerSources.ts:31]: Weird that this inherits properties like `visible` that it are not used by MapLayerSource? Is there really an inheritance relationship between them or just a coincidental similarity?

File: core/frontend/src/tile/map/MapLayerSources.ts [hunk 3]
```diff
@@ -25,16 +25,37 @@ export enum MapLayerSourceStatus {
...
+  public static fromJSON(json: MapLayerSourceProps): MapLayerSource | undefined {
+    if (json === undefined || json.url === undefined || undefined === json.name || undefined === json.formatId)
+      return undefined;
+
     const baseMap = json.baseMap === true;
-    return (typeof json.name === "string" && typeof json.url === "string" && typeof json.formatId === "string") ? new MapLayerSource(json.formatId, json.name, json.url, baseMap, json.transparentBackground === undefined ? true : json.transparentBackground, json.maxZoom, json.userName, json.password) : undefined;
+    return new MapLayerSource(json.formatId,
+      json.name,
+      json.url,
+      baseMap, json.transparentBackground === undefined ? true : json.transparentBackground,
```
Comment (pmconne) [inline core/frontend/src/tile/map/MapLayerSources.ts:57]: `false !== json.transparentBackground` would make it much easier to tell what's going on.

File: core/frontend/src/tile/map/MapLayerSources.ts [hunk 4]
```diff
@@ -25,16 +25,37 @@ export enum MapLayerSourceStatus {
...
+
 /** A source for map layers.  These may be catalogued for convenient use by users or applications.
  * @internal
  */
 export class MapLayerSource implements MapLayerProps {
   public subLayers?: MapSubLayerProps[];
 
-  private constructor(public formatId: string, public name: string, public url: string, public baseMap = false, public transparentBackground?: boolean, public maxZoom?: number, public userName?: string, public password?: string) { }
-  public static fromJSON(json: any): MapLayerSource | undefined {
+  private constructor(public formatId: string, public name: string, public url: string, public baseMap = false, public transparentBackground?: boolean, public userName?: string, public password?: string) { }
```
Comment (pmconne) [inline core/frontend/src/tile/map/MapLayerSources.ts:47]: nitpick but I hate these constructors that declare half a dozen class members inline - they are convenient once, for the author who gets to type a little less, and inconvenient forever, for everyone else who has to parse them.

**Review by pmconne — CHANGES_REQUESTED** [review]

File: core/frontend/src/tile/map/MapLayerSources.ts (hunk 4 above)
Comment (pmconne) [reply core/frontend/src/tile/map/MapLayerSources.ts:47]: Given the constructor is private and only called by `fromJson`, consider making it accept MapLayerSourceProps directly so you're not handling default values in two places.

**Comment by mdastous-bentley** [comment]
@pmconne Code updated based on your comments.  Are we good to go ?

**Review by pmconne — APPROVED** [review]

**Comment by pmconne** [comment]
> @pmconne Code updated based on your comments. Are we good to go ?

LGTM, gotta hassle the other required reviewers.

**Review by bsteinbk — APPROVED** [review]

**Pull request merged**